pkg/logs/logs.go

```diff
@@ -29,103 +29,114 @@ func NewLogsClient(kubernetesClient kclient.ClientInterface) *LogsClient {
 
 var _ Client = (*LogsClient)(nil)
 
-func (o *LogsClient) GetLogsForMode(mode string, componentName string, namespace string, follow bool) ([]map[string]io.ReadCloser, error) {
+func (o *LogsClient) GetLogsForMode(
+	mode string,
+	componentName string,
+	namespace string,
+	follow bool,
+	out io.Writer,
+	logChan chan map[string]interface{},
+	doneChan chan struct{},
+) error {
 	var selector string
-	logs := []map[string]io.ReadCloser{}
-	unknownMode := true
+	podChan := make(chan corev1.Pod) // grab the logs of the pod put on this channel
+	done := make(chan struct{})      // because populating doneChan directly would cause odo logs to exit prematurely.
+
+	go func() {
+		// this go routine gets the logs of the pods put on the podChan
+		for {
```

Comment:
When is this infinite loop finishing? I cannot see any return inside

Reply:
When the parent function populates `done` channel. It's implicit in nature rather than being explicit.

Reply:
This go routine has been removed. We now return a struct of channels from the business layer akin to what you recommended in https://github.com/redhat-developer/odo/pull/5942#issuecomment-1183394909.